snuba_sdk/__init__.py

```diff
@@ -0,0 +1,14 @@
+from abc import ABC, abstractmethod
+
+
+class Expression(ABC):
+    def __post_init__(self) -> None:
+        self.validate()
+
+    @abstractmethod
+    def validate(self) -> None:
+        raise NotImplementedError
```

Comment:
Are you sure an expression will be able to validate itself in isolation without context? 
Defining the method as part of the query structure means that either you provide context in the object to validate itself (which seems unlikely since they are immutable) or that validation is executed by each node independently without context.
This may or may not be possible depending on what you plan to validate. If you try to validate that a given column is in the logical schema, you need more context when calling validate on the column object.

Reply:
My plan is to let the expressions validate as much as possible independently, and then have some contextual validators at the query level. For example, a function can validate itself and then ensure its arguments are valid, but the query needs to assert that top level functions in the SELECT clause have aliases.